### PR TITLE
Fix bug where repositories are attached unnecessarily causing error logs

### PR DESCRIPTION
**Pull request by taimoorgit**
## Issues

https://github.com/OpsLevel/kubectl-opslevel/issues/286

https://github.com/OpsLevel/team-platform/issues/305

## Changelog

- [x] Edit `handleRepositories()` to only update a service repository if it needs changes, make logic more clear by using happy path, add more structured logging
- [x] Add unit testing for repository handling
  - [x] Case where repository does not exist in OpsLevel/get API call fails (skips)
  - [x] Case where repository is attached and needs no updates
  - [x] Case where repository is not attached and needs to be attached
  - [x] Case where repository is attached and needs updates
- [x] Make a `changie` entry
- [x] Add some TODO comments, fix 1 typo in test data (object)

## TODO: Tophatting

Deployment:

```
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
    opslevel.com/descriptionX: this was made for testing k8s sync
    opslevel.com/lifecycle: alpha
    opslevel.com/owner: platform
    opslevel.com/repo.helloworld1: github.com:taimoor-test-org/helloworld1
    opslevel.com/repo.helloworld2: github.com:taimoor-test-org/helloworld2
    opslevel.com/tags.from_kubernetes: "True"
    opslevel.com/tags.hello: world
    opslevel.com/tags.image: "1.24"
    opslevel.com/tags.is_public: "True"
    opslevel.com/tags.vpc_enabled: "True"
    opslevel.com/tier: tier_2
    repo: github.com:taimoor-test-org/hello_word
    repository: github.com:taimoor-test-org/hewwo_worlb
    ...
```

Config: (default sample config)

### Create new service, attach repositories, should succeed

```
~/r/kubectl-opslevel/src ta/handle-repo-mar-28 ❯ go run main.go service import -c opslevel-k8s.yaml --log-level=DEBUG           х INT Ruby 3.0.0 11:51:37 am
11:51AM DBG maxprocs: Leaving GOMAXPROCS=10: CPU quota undefined
11:51AM DBG Caching 'Tier' lookup table from API ...
11:51AM DBG Caching 'Lifecycle' lookup table from API ...
11:51AM DBG Caching 'Team' lookup table from API ...
11:51AM INF [apps/v1/deployments] Informer is ready and synced
11:51AM DBG mainloop: running from top queue_addr=0x14000674140 queue_len=3
11:51AM DBG mainloop: running from top queue_addr=0x14000674140 queue_len=2
11:51AM DBG mainloop: running from top queue_addr=0x14000674140 queue_len=1
11:51AM DBG object with key 'kube-system/coredns' skipped because it matches filter
11:51AM DBG mainloop: running from top queue_addr=0x14000674140 queue_len=0
11:51AM WRN unexpected happened: got service with alias 'default-demonstration' but the result has no ID
11:51AM WRN unexpected happened: got service with alias 'k8s:demonstration-default' but the result has no ID
11:51AM WRN [demonstration] Unable to find 'Team' with alias 'platform'
11:51AM INF [demonstration] Created new service
11:51AM INF [demonstration] Assigned alias 'default-demonstration'
11:51AM INF [demonstration] Assigned alias 'k8s:demonstration-default'
11:51AM INF [demonstration] Assigned tags: [{"key":"app","value":"nginx"},{"key":"from_kubernetes","value":"True"},{"key":"hello","value":"world"},{"key":"image","value":"1.24"},{"key":"imported","value":"kubectl-opslevel"},{"key":"is_public","value":"True"},{"key":"vpc_enabled","value":"True"}]
11:51AM INF successfully created a new service repository base_dir= repo=github.com:taimoor-test-org/helloworld1
11:51AM INF successfully created a new service repository base_dir= repo=github.com:taimoor-test-org/helloworld2
11:51AM INF successfully created a new service repository base_dir= repo=github.com:taimoor-test-org/hewwo_worlb
11:51AM DBG Breaking: on quit
11:51AM INF [demonstration-nomatch] No changes detected to fields - skipping update
11:51AM INF [demonstration-nomatch] All tags already assigned to service.
11:51AM INF [local-path-provisioner] No changes detected to fields - skipping update
11:51AM INF [local-path-provisioner] All tags already assigned to service.
11:51AM INF Import Complete
```

## Re-run after repositories are attached, should not see errors

```
~/r/kubectl-opslevel/src ta/handle-repo-mar-28 ❯ go run main.go service import -c opslevel-k8s.yaml --log-level=DEBUG              7s Ruby 3.0.0 11:51:48 am
11:52AM DBG maxprocs: Leaving GOMAXPROCS=10: CPU quota undefined
11:52AM DBG Caching 'Tier' lookup table from API ...
11:52AM DBG Caching 'Lifecycle' lookup table from API ...
11:52AM DBG Caching 'Team' lookup table from API ...
11:52AM INF [apps/v1/deployments] Informer is ready and synced
11:52AM DBG mainloop: running from top queue_addr=0x140004e8140 queue_len=3
11:52AM DBG mainloop: running from top queue_addr=0x140004e8140 queue_len=2
11:52AM DBG mainloop: running from top queue_addr=0x140004e8140 queue_len=1
11:52AM DBG object with key 'kube-system/coredns' skipped because it matches filter
11:52AM DBG mainloop: running from top queue_addr=0x140004e8140 queue_len=0
11:52AM WRN [demonstration] Unable to find 'Team' with alias 'platform'
11:52AM INF [demonstration] No changes detected to fields - skipping update
11:52AM INF [demonstration] All tags already assigned to service.
11:52AM DBG found service repository (Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjUyMDA) has display name: 'helloworld1' base_dir= repo=github.com:taimoor-test-org/helloworld1
11:52AM INF service repository (Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjUyMDA) does not require any updates base_dir= repo=github.com:taimoor-test-org/helloworld1
11:52AM DBG found service repository (Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjUyMDE) has display name: 'helloworld2' base_dir= repo=github.com:taimoor-test-org/helloworld2
11:52AM INF service repository (Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjUyMDE) does not require any updates base_dir= repo=github.com:taimoor-test-org/helloworld2
11:52AM DBG found service repository (Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjUyMDI) has display name: 'My Cool Repo' base_dir= repo=github.com:taimoor-test-org/hewwo_worlb
11:52AM INF service repository (Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjUyMDI) does not require any updates base_dir= repo=github.com:taimoor-test-org/hewwo_worlb
11:52AM DBG Breaking: on quit
11:52AM INF [demonstration-nomatch] No changes detected to fields - skipping update
11:52AM INF [demonstration-nomatch] All tags already assigned to service.
11:52AM INF [local-path-provisioner] No changes detected to fields - skipping update
11:52AM INF [local-path-provisioner] All tags already assigned to service.
11:52AM INF Import Complete
```

image

## Try changing display name

Modify config line to be:

```
                - '{"name": "Changed Display Name Here", "directory": "", "repo": .metadata.annotations.repository} | if .repo then . else empty end'
```

```
~/repos/kubectl-opslevel/src ta/handle-repo-mar-28 ❯ go run main.go service import -c opslevel-k8s.yaml --log-level=DEBUG                                                                                                     5s Ruby 3.0.0 12:02:09 pm
12:02PM DBG maxprocs: Leaving GOMAXPROCS=10: CPU quota undefined
12:02PM DBG Caching 'Tier' lookup table from API ...
12:02PM DBG Caching 'Lifecycle' lookup table from API ...
12:02PM DBG Caching 'Team' lookup table from API ...
12:02PM INF [apps/v1/deployments] Informer is ready and synced
12:02PM DBG mainloop: running from top queue_addr=0x140006bb100 queue_len=3
12:02PM DBG mainloop: running from top queue_addr=0x140006bb100 queue_len=2
12:02PM DBG mainloop: running from top queue_addr=0x140006bb100 queue_len=1
12:02PM DBG object with key 'kube-system/coredns' skipped because it matches filter
12:02PM DBG mainloop: running from top queue_addr=0x140006bb100 queue_len=0
12:02PM WRN [demonstration] Unable to find 'Team' with alias 'platform'
12:02PM INF [demonstration] No changes detected to fields - skipping update
12:02PM INF [demonstration] All tags already assigned to service.
12:02PM DBG found service repository (Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjUyMDI) has display name: 'My Cool Repo' base_dir= repo=github.com:taimoor-test-org/hewwo_worlb
12:02PM INF successfully updated service repository (Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjUyMDI) base_dir= repo=github.com:taimoor-test-org/hewwo_worlb
12:02PM DBG Breaking: on quit
12:02PM INF [demonstration-nomatch] No changes detected to fields - skipping update
12:02PM INF [demonstration-nomatch] All tags already assigned to service.
12:02PM INF [local-path-provisioner] No changes detected to fields - skipping update
12:02PM INF [local-path-provisioner] All tags already assigned to service.
12:02PM INF Import Complete
~/repos/kubectl-opslevel/src ta/handle-repo-mar-28 ❯ go run main.go service import -c opslevel-k8s.yaml --log-level=DEBUG                                                                                                     5s Ruby 3.0.0 12:02:50 pm
12:02PM DBG maxprocs: Leaving GOMAXPROCS=10: CPU quota undefined
12:02PM DBG Caching 'Tier' lookup table from API ...
12:02PM DBG Caching 'Lifecycle' lookup table from API ...
12:02PM DBG Caching 'Team' lookup table from API ...
12:02PM INF [apps/v1/deployments] Informer is ready and synced
12:02PM DBG mainloop: running from top queue_addr=0x14000881340 queue_len=3
12:02PM DBG mainloop: running from top queue_addr=0x14000881340 queue_len=2
12:02PM DBG mainloop: running from top queue_addr=0x14000881340 queue_len=1
12:02PM DBG object with key 'kube-system/coredns' skipped because it matches filter
12:02PM DBG mainloop: running from top queue_addr=0x14000881340 queue_len=0
12:02PM WRN [demonstration] Unable to find 'Team' with alias 'platform'
12:02PM INF [demonstration] No changes detected to fields - skipping update
12:02PM INF [demonstration] All tags already assigned to service.
12:02PM DBG found service repository (Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjUyMDI) has display name: 'Changed Display Name Here' base_dir= repo=github.com:taimoor-test-org/hewwo_worlb
12:02PM INF service repository (Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjUyMDI) does not require any updates base_dir= repo=github.com:taimoor-test-org/hewwo_worlb
12:02PM DBG Breaking: on quit
12:02PM INF [demonstration-nomatch] No changes detected to fields - skipping update
12:02PM INF [demonstration-nomatch] All tags already assigned to service.
12:02PM INF [local-path-provisioner] No changes detected to fields - skipping update
12:02PM INF [local-path-provisioner] All tags already assigned to service.
12:02PM INF Import Complete
```

## When directory (base directory) is not defined (added after realizing these were not being read on the kubectl end)

```
12:11PM WRN no base directory was associated with this repo - using root directory (/) repo=github.com:taimoor-test-org/hello_world service=demonstration
12:11PM INF successfully created a new service repository base_dir= repo=github.com:taimoor-test-org/hello_world service=demonstration
```

## Use customer's config from ticket

```
            repositories:
                - '.metadata.annotations | to_entries | map(select(.key | startswith("opslevel.com/repo"))) | map(.value)'

```

```
~/repos/kubectl-opslevel/src ta/handle-repo-mar-28 !1 ❯ k get deploy/demonstration -o json | jq '.metadata.annotations | to_entries | map(select(.key | startswith("opslevel.com/repo"))) | map(.value)'         21s Ruby 3.0.0 ○ kind-kind 12:19:52 pm
[
  "github.com:taimoor-test-org/helloworld1",
  "github.com:taimoor-test-org/helloworld2",
  "github.com:taimoor-test-org/repo7"
]
~/repos/kubectl-opslevel/src ta/handle-repo-mar-28 !1 ❯ go run main.go service import -c opslevel-k8s.yaml --log-level=DEBUG                                                                                                     Ruby 3.0.0 12:19:53 pm
12:20PM DBG maxprocs: Leaving GOMAXPROCS=10: CPU quota undefined
12:20PM DBG Caching 'Tier' lookup table from API ...
12:20PM DBG Caching 'Lifecycle' lookup table from API ...
12:20PM DBG Caching 'Team' lookup table from API ...
12:20PM INF [apps/v1/deployments] Informer is ready and synced
12:20PM DBG mainloop: running from top queue_addr=0x140001d3ca0 queue_len=3
12:20PM DBG mainloop: running from top queue_addr=0x140001d3ca0 queue_len=2
12:20PM DBG mainloop: running from top queue_addr=0x140001d3ca0 queue_len=1
12:20PM DBG object with key 'kube-system/coredns' skipped because it matches filter
12:20PM DBG mainloop: running from top queue_addr=0x140001d3ca0 queue_len=0
12:20PM WRN [demonstration] Unable to find 'Team' with alias 'platform'
12:20PM INF [demonstration] No changes detected to fields - skipping update
12:20PM INF [demonstration] All tags already assigned to service.
12:20PM WRN no base directory was associated with this repo - using root directory (/) repo=github.com:taimoor-test-org/helloworld1 service=demonstration
12:20PM DBG found service repository (Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjUyMDA) has display name: 'helloworld1' base_dir= repo=github.com:taimoor-test-org/helloworld1 service=demonstration
12:20PM INF service repository (Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjUyMDA) does not require any updates base_dir= repo=github.com:taimoor-test-org/helloworld1 service=demonstration
12:20PM WRN no base directory was associated with this repo - using root directory (/) repo=github.com:taimoor-test-org/helloworld2 service=demonstration
12:20PM DBG found service repository (Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjUyMDE) has display name: 'helloworld2' base_dir= repo=github.com:taimoor-test-org/helloworld2 service=demonstration
12:20PM INF service repository (Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjUyMDE) does not require any updates base_dir= repo=github.com:taimoor-test-org/helloworld2 service=demonstration
12:20PM WRN no base directory was associated with this repo - using root directory (/) repo=github.com:taimoor-test-org/repo7 service=demonstration
12:20PM INF successfully created a new service repository base_dir= repo=github.com:taimoor-test-org/repo7 service=demonstration
12:20PM DBG Breaking: on quit
12:20PM INF [demonstration-nomatch] No changes detected to fields - skipping update
12:20PM INF [demonstration-nomatch] All tags already assigned to service.
12:20PM INF [local-path-provisioner] No changes detected to fields - skipping update
12:20PM INF [local-path-provisioner] All tags already assigned to service.
12:20PM INF Import Complete

```

### CODE REVIEW

File: src/common/reconciler_test.go [hunk 1]
```diff
@@ -16,9 +16,8 @@ import (
 	"github.com/rocktavious/autopilot/v2023"
 )
 
-// TODO: tests here would be much easier to use if we had a builder for service registrations
-// TODO: tests here would be much easier to use if we had a builder for existing services
-// TODO: tests here would be much easier to use if we had a builder for clients so that we don't have to `panic("should not be called")` everything.
+// TODO: use builder pattern for client so that every function that isn't explicitly stubbed will panic
```
Comment (taimoorgit) [inline src/common/reconciler_test.go:19]: I added this TODO - we could probably cut the lines of code in this file by 40% by using a builder pattern.

I don't think a builder for service/service registration is necessary, so those lines are removed.

In progress: https://github.com/OpsLevel/kubectl-opslevel/pull/283

File: src/common/reconciler.go [hunk 2]
```diff
@@ -365,52 +365,62 @@ func (r *ServiceReconciler) handleTools(service *opslevel.Service, registration
...
 		}
-		repoCreateString := "repoCreate{}"
-		b, err := json.Marshal(repositoryCreate)
-		if err == nil {
-			repoCreateString = string(b)
-		}
-		foundRepository, foundRepositoryErr := r.client.GetRepositoryWithAlias(*repositoryCreate.Repository.Alias)
+
+		// setup logger to use for this repository
+		repoLogger := log.With().Str("repo", *inputRepository.Repository.Alias).Str("base_dir", *inputRepository.BaseDirectory).Logger()
```
[the remark this answers is not in the document]
Comment (rocktavious) [reply src/common/reconciler.go:375]: nice utilization of zero logger! ❤️

File: src/common/reconciler.go [hunk 3]
```diff
@@ -365,52 +365,62 @@ func (r *ServiceReconciler) handleTools(service *opslevel.Service, registration
...
+				continue
+			}
+			repoLogger.Info().Msgf("successfully updated service repository (%s)", serviceRepository.Id)
 			continue
 		}
-		repositoryCreate.Service = opslevel.IdentifierInput{Id: &service.Id}
-		err = r.client.CreateServiceRepository(repositoryCreate)
+
+		// if the ServiceRepository is not found, create the ServiceRepository (assign the repository to the current service)
+		inputRepository.Service = opslevel.IdentifierInput{Alias: &service.Aliases[0]}
```
Comment (rocktavious) [inline src/common/reconciler.go:417]: why did we switch away from ID here?

Comment (taimoorgit) [reply src/common/reconciler.go]: Pushed commit to use service ID here

**Review by rocktavious — APPROVED** [review]

**Review by rocktavious — APPROVED** [review]

**Pull request merged**